File: 649-Dota2Senate/649-Dota2Senate.py
from collections import deque 
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        r = deque([])
        d = deque([])


        #1- loop through the senate
        for i, s in enumerate(senate):    
            #2- separate the parties in two queues with their indexes
            r.append(i) if s == "R" else d.append(i)
    
        
        while r and d:
            #4- recalculate the index
            new_senate_r = r.popleft()
            new_senate_d = d.popleft()
                
            #5- compare the parties.
            #6- reappend the index of the party whith smaller index.
            if new_senate_r < new_senate_d:
                r.append(new_senate_r + len(senate))
            else:
                d.append(new_senate_d + len(senate))
                            
        
        #6- return the party with elements
        return 'Radiant' if r else 'Dire'


        

File: 649-Dota2Senate/649-Dota2Senate.js
/**
 * @param {string} senate
 * @return {string}
 */
var predictPartyVictory = function (senate) {

    let banRadiant = 0;
    let banDire = 0;
    let input = senate.split("");
    let hasRadiant = true;
    let hasDire = true;
    while (hasRadiant && hasDire) {
        hasRadiant = false;
        hasDire = false;

        for (i = 0; i < input.length; i++) {

            if (input[i] == 'R') {
                if (banRadiant == 0) {
                    hasRadiant = true;
                    banDire++

                } else {
                    banRadiant--
                    input[i] = 0;
                }
            } else if (input[i] == 'D') {
                if (banDire <= 0) {
                    hasDire = true;
                    //  Dire.push(true)
                    banRadiant++
                } else {
                    //banDire = Math.min(0,)
                    input[i] = 0;
                    banDire--
                }


            }
            else {
                continue;
            }


        }


    }
    return hasDire ? 'Dire' : 'Radiant'

};